core/src/test/java/org/elasticsearch/index/replication/ESIndexLevelReplicationTestCase.java

```diff
@@ -0,0 +1,478 @@
+                recoverReplica(replica, targetSupplier);
+                return null;
+            });
+            threadPool.generic().execute(task);
+            return task;
+        }
+
+        public synchronized void assertAllEqual(int expectedCount) throws IOException {
+            Set<Uid> primaryIds = getShardDocUIDs(primary);
+            assertThat(primaryIds.size(), equalTo(expectedCount));
+            for (IndexShard replica : replicas) {
+                Set<Uid> replicaIds = getShardDocUIDs(replica);
+                Set<Uid> temp = new HashSet<>(primaryIds);
+                temp.removeAll(replicaIds);
+                assertThat(replica.routingEntry() + " is missing docs", temp, empty());
+                temp = new HashSet<>(replicaIds);
+                temp.removeAll(primaryIds);
+                assertThat(replica.routingEntry() + " has extra docs", temp, empty());
+            }
+        }
+
+        private Set<Uid> getShardDocUIDs(final IndexShard shard) throws IOException {
+            shard.refresh("get_uids");
+            try (Engine.Searcher searcher = shard.acquireSearcher("test")) {
+                Set<Uid> ids = new HashSet<>();
+                searcher.searcher().search(new MatchAllDocsQuery(), new SimpleCollector() {
+                    private LeafReader reader;
+                    private FieldsVisitor fieldVisitor = new FieldsVisitor(false);
+
+                    @Override
+                    protected void doSetNextReader(LeafReaderContext context) throws IOException {
+                        reader = context.reader();
+                    }
+
+                    @Override
+                    public void collect(int doc) throws IOException {
+                        fieldVisitor.reset();
+                        reader.document(doc, fieldVisitor);
+                        ids.add(fieldVisitor.uid());
+                    }
+
+                    @Override
+                    public boolean needsScores() {
+                        return false;
+                    }
+                });
+                return ids;
+            }
+        }
+
+        public synchronized void refresh(String source) {
+            Stream.concat(replicas.stream(), Stream.of(primary)).forEach(s -> s.refresh(source));
+        }
+
+        public synchronized void flush() {
+            final FlushRequest request = new FlushRequest();
+            Stream.concat(replicas.stream(), Stream.of(primary)).forEach(s -> s.flush(request));
+        }
+
+        public synchronized List<ShardRouting> shardRoutings() {
+            return Stream.concat(replicas.stream(), Stream.of(primary)).map(IndexShard::routingEntry).collect(Collectors.toList());
+        }
+
+        @Override
+        public synchronized void close() throws Exception {
+            if (closed == false) {
+                closed = true;
+                for (IndexShard shard : Iterables.<IndexShard>concat(replicas, Collections.singletonList(primary))) {
```

Comment:
make sure you use IOUtils here - I wonder if we should implement Closeable on IndexShard and just forward to `close("API", false)` in there then we pass all of them in to a IOUtils.close call?